File: regions/system/library/typescript/distributed-lock/src/postgres-lock.ts
import { randomBytes } from 'crypto';
import type { Pool } from 'pg';
import { LockError, type LockGuard } from './index.js';

export class PostgresDistributedLock {
  private readonly keyPrefix: string;

  constructor(
    private readonly pool: Pool,
    keyPrefix = 'lock',
  ) {
    this.keyPrefix = keyPrefix;
  }

  private fullKey(key: string): string {
    return `${this.keyPrefix}:${key}`;
  }

  async acquire(key: string, _ttlMs: number): Promise<LockGuard> {
    const full = this.fullKey(key);
    const token = randomBytes(16).toString('hex');

    const result = await this.pool.query<{ pg_try_advisory_lock: boolean }>(
      'SELECT pg_try_advisory_lock(hashtext($1))',
      [full],
    );

    if (!result.rows[0].pg_try_advisory_lock) {
      throw new LockError(`lock already held: ${key}`, 'LOCK_HELD');
    }

    return { key, token };
  }

  async release(guard: LockGuard): Promise<void> {
    const full = this.fullKey(guard.key);

    const result = await this.pool.query<{ pg_advisory_unlock: boolean }>(
      'SELECT pg_advisory_unlock(hashtext($1))',
      [full],
    );

    if (!result.rows[0].pg_advisory_unlock) {
      throw new LockError(`lock not owned: ${guard.key}`, 'NOT_OWNER');
    }
  }

  async isLocked(key: string): Promise<boolean> {
    const full = this.fullKey(key);

    const result = await this.pool.query<{ exists: boolean }>(
      "SELECT EXISTS(SELECT 1 FROM pg_locks WHERE locktype = 'advisory' AND classid = hashtext($1)::int) AS exists",
      [full],
    );

    return result.rows[0].exists;
  }
}

File: regions/system/library/dart/distributed_lock/lib/src/postgres_lock.dart
import 'dart:math';

import 'package:postgres/postgres.dart';

import 'lock.dart';

class PostgresDistributedLock implements DistributedLock {
  final Connection _conn;
  final _rng = Random.secure();

  PostgresDistributedLock(this._conn);

  String _generateToken() {
    return List.generate(
            16, (_) => _rng.nextInt(256).toRadixString(16).padLeft(2, '0'))
        .join();
  }

  /// Returns the advisory lock key as an integer hash of the key string.
  /// Uses hashtext equivalent: we use Dart's string hash and mask to 32-bit.
  int _lockKey(String key) => key.hashCode & 0x7fffffff;

  @override
  Future<LockGuard> acquire(String key, Duration ttl) async {
    final lockKey = _lockKey(key);
    final rows = await _conn.execute(
      Sql.named('SELECT pg_try_advisory_lock(@key)'),
      parameters: {'key': lockKey},
    );
    final acquired = rows.first[0] as bool? ?? false;
    if (!acquired) {
      throw LockException('lock already held for key: $key');
    }
    final token = _generateToken();
    return LockGuard(key: key, token: token);
  }

  @override
  Future<void> release(LockGuard guard) async {
    final lockKey = _lockKey(guard.key);
    final rows = await _conn.execute(
      Sql.named('SELECT pg_advisory_unlock(@key)'),
      parameters: {'key': lockKey},
    );
    final released = rows.first[0] as bool? ?? false;
    if (!released) {
      throw LockException(
          'lock not held or token mismatch for key: ${guard.key}');
    }
  }

  @override
  Future<bool> isLocked(String key) async {
    final lockKey = _lockKey(key);
    final rows = await _conn.execute(
      Sql.named(
          'SELECT COUNT(*) FROM pg_locks WHERE locktype = \'advisory\' AND objid = @key AND granted = true'),
      parameters: {'key': lockKey},
    );
    final count = rows.first[0] as int? ?? 0;
    return count > 0;
  }
}
